[FSDP] Do not clean FQNs even for `use_orig_params=True`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #92035 [FSDP][RFC] Enforce rank `r`'s current device is `cuda:r`
* **#91767 [FSDP] Do not clean FQNs even for `use_orig_params=True`**
* #92031 [FSDP][BE] Improve `device_id` + CPU offload test
* #92028 [FSDP][BE] Rename `prefixed_param_names` -> `fqns` for consolidation
* #92027 [FSDP][BE] Better error msg for incorrect device for training

Cleaning FQN for `FullyShardedDataParallel(use_orig_params=True)` can cause some discrepancies with respect to the FQN compared to manually looping over `named_modules()` and `named_parameters()` together.

There is no requirement for the FQNs to be clean when using wrapper FSDP + `use_orig_params=True`. We can leave clean FQNs to `fully_shard`.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

Commit:
[WIP] Do not clean FQNs even for `use_orig_params=True`

[ghstack-poisoned]

Commit:
Update on "[WIP] Do not clean FQNs even for `use_orig_params=True`"




[ghstack-poisoned]

Commit:
Update on "[WIP] Do not clean FQNs even for `use_orig_params=True`"


Cleaning FQN for `FullyShardedDataParallel(use_orig_params=True)` can cause some discrepancies with respect to the FQN compared to manually looping over `named_modules()` and `named_parameters()` together.

Testing in CI...

[ghstack-poisoned]

Commit:
Update on "[WIP] Do not clean FQNs even for `use_orig_params=True`"


Cleaning FQN for `FullyShardedDataParallel(use_orig_params=True)` can cause some discrepancies with respect to the FQN compared to manually looping over `named_modules()` and `named_parameters()` together.

Testing in CI...

[ghstack-poisoned]

Commit:
Update on "[FSDP] Do not clean FQNs even for `use_orig_params=True`"


Cleaning FQN for `FullyShardedDataParallel(use_orig_params=True)` can cause some discrepancies with respect to the FQN compared to manually looping over `named_modules()` and `named_parameters()` together.

Testing in CI...

[ghstack-poisoned]

Commit:
Update on "[FSDP] Do not clean FQNs even for `use_orig_params=True`"


Cleaning FQN for `FullyShardedDataParallel(use_orig_params=True)` can cause some discrepancies with respect to the FQN compared to manually looping over `named_modules()` and `named_parameters()` together.

Testing in CI...

[ghstack-poisoned]

huydhn commented:
@pytorchbot revert -m 'This breaks inductor_distributed workflow https://hud.pytorch.org/pytorch/pytorch/commit/a383789f4d8ecb36adaff6bd3746430209ff0546' -c nosignal

Commit:
Update on "[FSDP] Do not clean FQNs even for `use_orig_params=True`"


Cleaning FQN for `FullyShardedDataParallel(use_orig_params=True)` can cause some discrepancies with respect to the FQN compared to manually looping over `named_modules()` and `named_parameters()` together.

There is no requirement for the FQNs to be clean when using wrapper FSDP + `use_orig_params=True`. We can leave clean FQNs to `fully_shard`.

[ghstack-poisoned]

awgu commented on torch/_dynamo/testing.py:
```diff
@@ -46,7 +46,7 @@ def remove_optimized_module_prefix(name):
     assert name.startswith(prefix)
     name = name[len(prefix) :]
-    return torch.distributed.fsdp._common_utils.clean_tensor_name(name)
+    return name
```
@wconstab

Context: This PR changes FSDP so that when `use_orig_params=True`, the parameter names returned from `named_parameters()` include the `_fsdp_wrapped_module.` prefixes. Before, FSDP overrode `named_parameters()` to clean the prefix (see `fully_sharded_data_parallel.py`). I found that overriding it to clean the prefix creates unwanted discrepancy with the true module structure, which can lead to some issues. We can leave having clean parameter names to the currently-developed composable APIs (e.g. `fully_shard`).

This change here in `_dynamo/testing.py` to not call `clean_tensor_name()` allows the unit tests to pass. I still wanted to check: does Dynamo + FSDP rely on FSDP's `named_parameters()` returning the cleaned names (i.e. without `_fsdp_wrapped_module.`)?

wconstab replied:
I don't think dynamo relies on the structure of the name either way.  other than we do rely on finding the "is_fsdp_wrapped_module" boolean for enablement of fsdp handling.


question: what is the history of `remove_optimized_module_prefix` fn?  at a glance, i have no memory of modifying this myself for fsdp but i see it had an fsdp clean param thing in it.  did you put that there before and are taking it out now?

awgu replied:
`removed_optimized_module_prefix` came from https://github.com/pytorch/pytorch/pull/89113

My understanding is as follows:
https://github.com/pytorch/pytorch/blob/46a81c8db786ac03abc16bbecc8401dcbf451f10/torch/_dynamo/testing.py#L63-L65
https://github.com/pytorch/pytorch/blob/46a81c8db786ac03abc16bbecc8401dcbf451f10/test/distributed/test_dynamo_distributed.py#L368-L370

The `eager_model` removed the prefixes but the `opt_model` did not, so for parity, we needed to manually remove the prefixes for `opt_model`. However, with this PR, `eager_model` no longer removes the prefixes, so similarly, we can remove the manual removal for `opt_model`.

In this case, we should be safe to land this.

wconstab replied:
seems ok to me

Commit:
Update on "[FSDP] Do not clean FQNs even for `use_orig_params=True`"


Cleaning FQN for `FullyShardedDataParallel(use_orig_params=True)` can cause some discrepancies with respect to the FQN compared to manually looping over `named_modules()` and `named_parameters()` together.

There is no requirement for the FQNs to be clean when using wrapper FSDP + `use_orig_params=True`. We can leave clean FQNs to `fully_shard`.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

[ghstack-poisoned]

awgu commented:
@pytorchbot merge

malfet commented:
@pytorchbot revert -m "Looks like it broke `test_compatible_with_named_optimizer` distribued tests, see https://hud.pytorch.org/pytorch/pytorch/commit/d6f3265e1add26abedb504910be93b393b9fb33c" -c nosignal

awgu commented:
We had a land race. We will re-land this PR after a fix on the optimizer state side.
